0020-py-version-support.md

```diff
@@ -0,0 +1,251 @@
+
+Arch
+----
+Arch is a rolling release and is consistantly updating their Python version, so this is not
+of any concern for this policy.
+
+FreeBSD
+-------
+According to this https://wiki.freebsd.org/Python/PortsPolicy FreeBSD currently already aligns
+with this policy. They only support Python versions until they are End of Life so there are
+no concerns with this operating system.
+
+SmartOS
+-------
+SmartOS includes the latest Python version so will not be impacted by this change.
+
+
+What about dependencies?
+------------------------
+Before dropping an End of Life Python version we should have an idea beforehand if Salt's dependencies
+do not support a particular Python version with our testing suite. Currently our tests use the dependencies
+in the requirements directory in the main Salt project. This does of course mean that if we are not testing
+a module we would not have this information beforehand. We are now requiring tests to be added
+for PRs so this will help ensure this is caught for those modules, but there are still some modules that do
+not have tests currently.
+
+For the Sodium Release, I installed the libraries from anything defined with :depends: in the documentation
+and the requirements defined in our requirements directory. Once these were installed I ran
+this tool https://pypi.org/project/checkmyreqs/ against all installed Python libraries. This brought to light
+the following libraries that will not support Python 3, which will impact some Salt components:
+
+ - pycassa (salt/modules/cassandra.py, salt/returners/cassandra_return.py)
+ - vbox (salt/modules/vbox_guest.py, salt/modules/vbox_guest.py,
+         salt/modules/vboxmanage.py, salt/utils/virtualbox.py,
+         salt/cloud/clouds/virtualbox.py)
+
+Both of these projects have not had commits added to the project in years. This will need to be documented
+for the Sodium release to make it clear that these modules will not work unless the upstream projects add
+Python 3 support.
+
+Going forward adding more test coverage will help to uncover these projects before we drop a Python version,
+but if we do not have enough test coverage, this exercise should be done any time a Python Version is dropped.
+There might be a conflict between a Salt dependency and its Python version support that will impact many users.
+If this conflict does occur we may need to delay dropping support for the End of Life Python version until that
+dependency is updated to the appropriate Python version.
+
+## Who is Affected
+
+Salt Contributors
+-----------------
+When a Python Version is dropped any contributions will need to ensure they align with the syntax
+of the lowest supported Python version. These contributors can look at Python's Release Notes for
+details around what is added into a release or removed. For example for Python 3.6
+https://docs.Python.org/3/whatsnew/3.6.html outlines some of the changes and removals added to this version.
```

Comment:
> ensure they work with the lowest supported Python version. These contributors can look at Python's Release Notes for details of what has been removed or deprecated. For example https://docs.Python.org/3/whatsnew/3.6.html#removed lists what was removed in Python 3.6.

Most changes are not syntax. No "alignment" is required, and new features aren't relevant.

I notice that this fails to account for existing code that might break. Unless you require all contributors to maintain their contributions indefinitely?

Reply:
Are new contributions still allowed to use deprecated features? What's the timeline for updating new deprecations that arise in existing code? Will there be an automated system to capture them (e.g. from CI/integration tests)?

Reply:
I can just remove the word "syntax" and ensure it states contributions need to align with the lowest supported python version. And yes a contributor would need to align with both removals and additions.

New features are relevant. For example, if we begin to support >=Python 3.6 then a contributor can now use fstrings. Whereas when we support  >=python3.5 we cannot accept contributions with fstrings.

You cannot use deprecated features for your contributions. For example if we start supporting >=Python 3.6 you cannot use this deprecated code anymore:

`traceback.Ignore class and traceback.usage, traceback.modname, traceback.fullmodname, traceback.find_lines_from_code, traceback.find_lines, traceback.find_strings, traceback.find_executable_lines methods were removed from the traceback module. They were undocumented methods deprecated since Python 3.2 and equivalent functionality is available from private methods.`

For updating deprecations/testing I specifically call out when we add new python versions here:

https://github.com/saltstack/salt-enhancement-proposals/pull/26/files#diff-d745c39ab3665fe44caa19ff2a6dffdaR62

So at the beginning of each feature development cycle we need to see what python versions are being not only dropped but added and ensure testing is added around that version which should capture the deprecations.

Reply:
There's a big difference between "can use fstrings" and "must use fstrings". To "align with additions" means the latter, which IMO is unreasonable.

Reply:
yeah i agree with you that its not a must for new features, I can lighten up the language there.